### PR TITLE
Simplify comparator implementation

### PR DESCRIPTION
**Pull request by RoiEXLab**
Follow up to #2790
@DanVanAtta it seems like I forgot a closing curly bracket in my proposal ^^
Adding it back and everything worked fine

### CODE REVIEW

**Comment by RoiEXLab** [comment]
I added some characterization tests.
A potential issue with using classes of the standard library could be that this could break tests on a major update

File: src/test/java/games/strategy/engine/message/RemoteInterfaceHelperTest.java [hunk 1]
```diff
@@ -1,19 +1,43 @@
...
+        RemoteInterfaceHelper.getMethod(28, PrintStream.class));
+    checkMethodMatches("println", new Class<?>[] {boolean.class},
+        RemoteInterfaceHelper.getMethod(29, PrintStream.class));
+    checkMethodMatches("println", new Class<?>[] {char.class}, RemoteInterfaceHelper.getMethod(30, PrintStream.class));
+
+    assertEquals(27, RemoteInterfaceHelper.getNumber("println", new Class<?>[] {}, PrintStream.class));
+    assertEquals(34, RemoteInterfaceHelper.getNumber("println", new Class<?>[] {Object.class}, PrintStream.class));
+  }
+
+  private void checkMethodMatches(final String name, final Class<?>[] parameterTypes, final Method method) {
```
Comment (ssoloff) [inline src/test/java/games/strategy/engine/message/RemoteInterfaceHelperTest.java]: Nit: This method can be static.

**Comment by RoiEXLab** [comment]
@ssoloff I was actually suprised it worked right off the bat 😄

**Comment by ssoloff** [comment]
> A potential issue with using classes of the standard library could be that this could break tests on a major update

I think it's ok, especially given that we know from Travis that it works on both Java 8 and 9.

If you're really concerned, it would be easy to write a test-specific interface or class that would be guaranteed to not change outside of our control.

**Comment by RoiEXLab** [comment]
@ssoloff 
> I think it's ok, especially given that we know from Travis that it works on both Java 8 and 9.

> If you're really concerned, it would be easy to write a test-specific interface or class that would be guaranteed to not change outside of our control.

I think so as well. I thought about creatig a dummy class, but simply didn't do it, because it really affects test code only.
If it breaks one day, we could simply use another class that hasn't changed.

**Comment by ssoloff** [comment]
Just a note about test coverage...  Running the new tests on the old code shows that the `null` parameter type cases are not covered.  I don't know if that's important.  I can't see how `Method#getParameterTypes()` can ever return `null`, so additional test cases probably aren't necessary.

**Comment by RoiEXLab** [comment]
@ssoloff 
> Just a note about test coverage...  Running the new tests on the old code shows that the `null` parameter type cases are not covered.  I don't know if that's important.  I can't see how `Method#getParameterTypes()` can ever return `null`, so additional test cases probably aren't necessary.

This was intentional. My original approach was actually checking for null (see #2790), until I checked the javadocs that stated that the return value is an array, consisting of 0 elements when there are no arguments to pass.

File: src/main/java/games/strategy/engine/message/RemoteInterfaceHelper.java [hunk 2]
```diff
@@ -19,41 +17,26 @@ static int getNumber(final String methodName, final Class<?>[] argTypes, final C
...
-        return t1[i].getName().compareTo(t2[i].getName());
-      }
-    }
-    return 0;
-  };
+  private static final Comparator<Method> methodComparator = Comparator
+      .comparing(Method::getName)
+      .thenComparing(Method::getParameterTypes,
+          Comparator.comparingInt((Class<?>[] a) -> a.length)
+              .thenComparing((o1, o2) -> {
```
Comment (RoiEXLab) [inline src/main/java/games/strategy/engine/message/RemoteInterfaceHelper.java]: Note for the future: In java 9 this code can be replaced with https://docs.oracle.com/javase/9/docs/api/java/util/Arrays.html#compare-T:A-T:A-java.util.Comparator-
e.g.:
```java
private static final Comparator<Method> methodComparator = Comparator
  .comparing(Method::getName)
  .thenComparing(Method::getParameterTypes,
    (o1, o2) -> Arrays.compare(o1, o2, Comparator.comparing(Class::getName)));
```

**Pull request merged**